flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateAssignmentOperation.java

```diff
@@ -136,19 +136,24 @@ public void assignStates() {
 
         // repartition state
         for (TaskStateAssignment stateAssignment : vertexAssignments.values()) {
-            if (stateAssignment.hasNonFinishedState) {
+            if (stateAssignment.hasNonFinishedState
+                    // FLINK-31963: We need to run repartitioning for stateless operators that have
+                    // upstream output or downstream input states.
+                    || stateAssignment.hasUpstreamOutputStates()
+                    || stateAssignment.hasDownstreamInputStates()) {
                 assignAttemptState(stateAssignment);
             }
         }
 
         // actually assign the state
         for (TaskStateAssignment stateAssignment : vertexAssignments.values()) {
-            // If upstream has output states, even the empty task state should be assigned for the
-            // current task in order to notify this task that the old states will send to it which
-            // likely should be filtered.
+            // If upstream has output states or downstream has input states, even the empty task
+            // state should be assigned for the current task in order to notify this task that the
+            // old states will send to it which likely should be filtered.
             if (stateAssignment.hasNonFinishedState
                     || stateAssignment.isFullyFinished
-                    || stateAssignment.hasUpstreamOutputStates()) {
+                    || stateAssignment.hasUpstreamOutputStates()
+                    || stateAssignment.hasDownstreamInputStates()) {
```

Comment:
Should we compute once `hasUpstreamOutputStates` and `hasDownstreamInputStates` similar to `isFullyFinished`?
(that would at least be consistent, and also faster)

Reply:
I was considering that, and then I saw that most of the heavy work for the functions is already cached. Do you think it's still worth it?

Reply:
I think it is, it's still an iteration over all upstream subtasks for every downstream subtask in the worst case, and it would be more consistent with the other checks. 
But I'm also fine with the current way.